### PR TITLE
DETR implementation update

### PR DESCRIPTION
**Pull request by gunho1123**
# Description

> :memo: Please include a summary of the change. 
> Current DETR implementation supports TFDS only as an input source. To utilize my dataset or downloaded tfreocrd files in the framework, I extend it to support tfrecord inputs on local or cloud storage.
> In addition, I add a backbone factory in modeling to make flexible choosing backbone for feature encoding.
> I have organized the configuration list, which currently contains all the parameters in task configuration. I add subgroups for the parameters to make it clear. (task.model, task.losses, ...)
> No additional dependencies are required.
> TPU compatibility checked.

## Type of change

For a new feature or function, please create an issue first to discuss it
with us before submitting a pull request.

Note: Please delete options that are not relevant.

- [O] New feature (non-breaking change which adds functionality)

## Tests

> :memo: Please describe the tests that you ran to verify your changes.
>  
> * Provide instructions so we can reproduce.  
> * Please also list any relevant details for your test configuration.  

**Test Configuration**:

## Checklist

- [O] I have signed the [Contributor License Agreement](https://github.com/tensorflow/models/wiki/Contributor-License-Agreements).
- [O] I have read [guidelines for pull request](https://github.com/tensorflow/models/wiki/Submitting-a-pull-request).
- [O] My code follows the [coding guidelines](https://github.com/tensorflow/models/wiki/Coding-guidelines).
- [O] I have performed a self [code review](https://github.com/tensorflow/models/wiki/Code-review) of my own code.
- [O] I have commented my code, particularly in hard-to-understand areas.
- [O] I have made corresponding changes to the documentation.
- [O] My changes generate no new warnings.
- [O] I have added tests that prove my fix is effective or that my feature works.

### CODE REVIEW

**Comment by frederick0329** [comment]
High level comments, 
can we not delete coco.py and make sure both path works in official/projects/detr/configs/detr.py.
In official/projects/detr/configs/detr.py, you create another experiment using the tfrecord path.
We can then deprecate the coco.py path later.

**Comment by gunho1123** [comment]
@frederick0329 Thank you for your comments! I've created another experiment using the tfrecord path named 'detr_coco_tfrecord'. In addition, I've added "file_type" in COCODataConfig for "input_reader_factory.input_reader_generator". The problem is that it seems like the class label in TFDS is 0-based (starts from 0) whereas the class label in my tfrecord is 1-based. Would it be better to make a condition when TFDS is used? I've created my tfrecord using this script. (https://github.com/tensorflow/models/blob/master/official/vision/data/create_coco_tf_record.py)

File: official/projects/detr/tasks/detection.py [hunk 1]
```diff
@@ -37,46 +44,106 @@ class DectectionTask(base_task.Task):
...
+  """def build_inputs(self,
+                   params: detr_cfg.DataConfig,
+                   input_context: Optional[tf.distribute.InputContext] = None):
+    return coco.COCODataLoader(params).load(input_context)"""
+  
+  def build_inputs(self,
+                   params,
+                   input_context: Optional[tf.distribute.InputContext] = None):
     """Build input dataset."""
-    return coco.COCODataLoader(params).load(input_context)
```
Comment (frederick0329) [inline official/projects/detr/tasks/detection.py:56]: should we add a if statement
if params is COCODataConfig:
  return coco.COCODataLoader(params).load(input_context)

so we keep both path working and deprecate it later.

**Comment by frederick0329** [comment]
> @frederick0329 Thank you for your comments! I've created another experiment using the tfrecord path named 'detr_coco_tfrecord'. In addition, I've added "file_type" in COCODataConfig for "input_reader_factory.input_reader_generator". The problem is that it seems like the class label in TFDS is 0-based (starts from 0) whereas the class label in my tfrecord is 1-based. Would it be better to make a condition when TFDS is used? I've created my tfrecord using this script. (https://github.com/tensorflow/models/blob/master/official/vision/data/create_coco_tf_record.py)

I see now why you want to delete coco.py path.
You can add a flag in DETRTask where class labels starts and make the corrsponding changes.

**Comment by gunho1123** [comment]
@frederick0329 Thank you. That's a good idea.
So can I delete coco.py path? I can keep both paths working using the if statement as you mentioned above.

**Comment by frederick0329** [comment]
No, let's keep it for now (there are internal code depending on it). In this PR, let's make sure the code works with both detr_input.py and coco.py works.

File: official/projects/detr/configs/detr.py [hunk 2]
```diff
@@ -15,33 +15,82 @@
...
+from official.modeling import hyperparams
+from official.vision.configs import common
+from official.vision.configs import backbones
 from official.projects.detr import optimization
 from official.projects.detr.dataloaders import coco
 
+# pylint: disable=missing-class-docstring
+# Keep for backward compatibility.
+@dataclasses.dataclass
+class TfExampleDecoder(common.TfExampleDecoder):
```
Comment (frederick0329) [inline official/projects/detr/configs/detr.py:32]: where is this used?

**Comment by frederick0329** [comment]
Thank you so much!

File: official/projects/detr/configs/detr.py (hunk 2 above)
Comment (gunho1123) [reply official/projects/detr/configs/detr.py:32]: We can delete it. It seems useless.

File: official/projects/detr/configs/detr.py [hunk 3]
```diff
@@ -101,3 +157,140 @@ def detr_coco() -> cfg.ExperimentConfig:
...
+  decay_at = train_steps - 100 * steps_per_epoch  # 200 epochs
+  config = cfg.ExperimentConfig(
+      task=DetrTask(
+          init_checkpoint='gs://tf_model_garden/vision/resnet50_imagenet/ckpt-62400',
+          init_checkpoint_modules='backbone',
+          annotation_file=os.path.join(COCO_INPUT_PATH_BASE,
+                                       'instances_val2017.json'),
+          model=Detr(
+              input_size=[1333, 1333, 3],
+              norm_activation=common.NormActivation(use_sync_bn=False)),
```
Comment (frederick0329) [inline official/projects/detr/configs/detr.py:181]: shhould use_sync_bn default to True?

Comment (gunho1123) [reply official/projects/detr/configs/detr.py:181]: I don't think so. I will change it.

Comment (gunho1123) [reply official/projects/detr/configs/detr.py:164]: The default is defined in the "official.vision.configs import common".
I think it is out of my hands.

Comment (frederick0329) [reply official/projects/detr/configs/detr.py:164]: No, I mean should this line be
norm_activation=common.NormActivation(use_sync_bn=True)) ?
I think in DETR's original setup it is set to True.

Comment (frederick0329) [reply official/projects/detr/configs/detr.py:164]: nvm, I think we freeze the backbone.
https://github.com/tensorflow/models/blob/d81c7d4b35404bae2a7f3583668fdd1cfc51fce5/official/projects/detr/modeling/detr.py#L119

Comment (gunho1123) [reply official/projects/detr/configs/detr.py]: Thank you for the review. I changed the use_sync_bn to True. :)

Comment (frederick0329) [reply official/projects/detr/configs/detr.py]: We might need to add a field ```bn_trainable```  here?
https://github.com/tensorflow/models/blob/d81c7d4b35404bae2a7f3583668fdd1cfc51fce5/official/vision/configs/common.py#L113

Comment (gunho1123) [reply official/projects/detr/configs/detr.py]: Can I modify the common file? If I can, I will add the argument.

Comment (frederick0329) [reply official/projects/detr/configs/detr.py]: Never mind. I read the code more carefully, you don't need to make any changes, setting sync_batch_norm to True looks good as it shouldn't matter.

https://github.com/tensorflow/models/blob/d81c7d4b35404bae2a7f3583668fdd1cfc51fce5/official/vision/modeling/backbones/resnet.py#L432

File: official/projects/detr/configs/detr.py [hunk 4]
```diff
@@ -15,33 +15,65 @@
...
+# pylint: disable=missing-class-docstring
+@dataclasses.dataclass
+class DataConfig(cfg.DataConfig):
+  """Input config for training."""
+  input_path: str = ''
+  global_batch_size: int = 0
+  is_training: bool = False
+  dtype: str = 'bfloat16'
+  decoder: common.DataDecoder = common.DataDecoder()
+  #parser: Parser = Parser()
```
Comment (frederick0329) [inline official/projects/detr/configs/detr.py:38]: remove

File: official/projects/detr/configs/detr.py [hunk 5]
```diff
@@ -15,33 +15,65 @@
...
+
 from official.core import config_definitions as cfg
 from official.core import exp_factory
+from official.modeling import hyperparams
+from official.vision.configs import common
+from official.vision.configs import backbones
 from official.projects.detr import optimization
 from official.projects.detr.dataloaders import coco
 
+# pylint: disable=missing-class-docstring
```
Comment (frederick0329) [inline official/projects/detr/configs/detr.py:29]: let's try not adding this and add docstring is possible.

**Review by frederick0329 — COMMENTED** [review]
Last comments!

**Comment by gunho1123** [comment]
> Last comments!

Thank you for your valuable comments. If you have trouble reproducing the results, please let me know. :)

File: official/projects/detr/modeling/detr.py [hunk 6]
```diff
@@ -116,7 +116,9 @@ def __init__(self, num_queries, hidden_size, num_classes,
       raise ValueError("hidden_size must be a multiple of 2.")
     # TODO(frederickliu): Consider using the backbone factory.
     # TODO(frederickliu): Add to factory once we get skeleton code in.
-    self._backbone = resnet.ResNet(50, bn_trainable=False)
+    #self._backbone = resnet.ResNet(101, bn_trainable=False)
+    # (gunho) use backbone factory
```
Comment (frederick0329) [inline official/projects/detr/modeling/detr.py:120]: remove?

Comment (gunho1123) [reply official/projects/detr/modeling/detr.py:120]: Yes. my bad. I will delete it.

**Review by frederick0329 — APPROVED** [review]
Saw an extra comment that can be removed. Thank you for the PR!
